test/common/network/connection_impl_test.cc

```diff
@@ -67,47 +71,84 @@ TEST_P(ConnectionImplDeathTest, BadFd) {
                ".*assert failure: fd_ != -1.*");
 }
 
-class ConnectionImplTest : public testing::TestWithParam<Address::IpVersion> {};
+class ConnectionImplTest : public testing::TestWithParam<Address::IpVersion> {
+public:
+  void setUpBasicConnection() {
+    listener_ =
+        dispatcher_.createListener(connection_handler_, socket_, listener_callbacks_, stats_store_,
+                                   Network::ListenerOptions::listenerOptionsWithBindToPort());
+
+    client_connection_ = dispatcher_.createClientConnection(socket_.localAddress());
+    client_connection_->addConnectionCallbacks(client_callbacks_);
+  }
+
+  void connect() {
+    client_connection_->connect();
+    read_filter_.reset(new NiceMock<MockReadFilter>());
+    EXPECT_CALL(listener_callbacks_, onNewConnection_(_))
+        .WillOnce(Invoke([&](Network::ConnectionPtr& conn) -> void {
+          server_connection_ = std::move(conn);
+          server_connection_->addConnectionCallbacks(server_callbacks_);
+          server_connection_->addReadFilter(read_filter_);
+        }));
+    EXPECT_CALL(client_callbacks_, onEvent(ConnectionEvent::Connected));
+    dispatcher_.run(Event::Dispatcher::RunType::NonBlock);
+  }
+
+  void disconnect() {
+    EXPECT_CALL(client_callbacks_, onEvent(ConnectionEvent::LocalClose));
+    client_connection_->close(ConnectionCloseType::NoFlush);
+    dispatcher_.run(Event::Dispatcher::RunType::NonBlock);
+  }
+
+  void useMockBuffer() {
+    dynamic_cast<ConnectionImpl*>(client_connection_.get())
+        ->replaceWriteBufferForTest(std::move(buffer_ptr_));
+  }
+
+protected:
+  Event::DispatcherImpl dispatcher_;
+  Stats::IsolatedStoreImpl stats_store_;
+  Network::TcpListenSocket socket_{Network::Test::getAnyAddress(GetParam()), true};
+  Network::MockListenerCallbacks listener_callbacks_;
+  Network::MockConnectionHandler connection_handler_;
+  Network::ListenerPtr listener_;
+  Network::ClientConnectionPtr client_connection_;
+  MockConnectionCallbacks client_callbacks_;
+  Network::ConnectionPtr server_connection_;
+  Network::MockConnectionCallbacks server_callbacks_;
+  std::shared_ptr<MockReadFilter> read_filter_;
+  std::unique_ptr<MockBuffer> buffer_ptr_{new MockBuffer()};
+  MockBuffer& buffer_{*buffer_ptr_};
```

Comment:
You can make this an actual pointer as per earlier thread and idiom discussion if you want.